### ISSUE TITLE
Introduce a new level of indirection to the TokenMap to further optimize token removal by doing only a single memory operation

### ISSUE THREAD

**Issue by kayagokalp** [FuelLabs/sway#6482]
This is all based on the understanding that I have how LSP TokenMap works as of now, so I am expecting it to have lots of holes around it, sorry if this doesn't even make sense at the first place 🙌 

We can probably change how we represent the token map, to hold references to the smallest unit that we know will be effected once the user starts to type.  With this PR this "smallest" unit becomes a file. So TokenMap would be structured as a graph, where leaf nodes are tokens and the parent nodes are the files owning those tokens. Once you type something, you would directly kill the parent node that represents that file, rather than removing each element for a given file, one by one for each keystroke. (i.e mark parent file node invalid and in next read re-create it or if we have garbage collection, it can directly re-populate that section of the graph, might be lighter than actually doing a memory operation to delete them, 1 mem operation compared N operation where N is the number of Tokens we have in the modified file to delete the invalid sections in token map).

Also, paging them by the file, gives us the opportunity to lay out the actual tokens sequentially in memory. Which might have better cache profile and lead to faster traversal times. So we would have graph, that has nodes that represents each file, those parents would poin to a memory location, and all tokens would be sequential for that file from that location. Havin said that, I am not sure LSP token "read characteristics" have spatial and/or temporal locality. So this might not gain us anything in the end.
Also this creates another possibility after the  "file" node, we can have "region" nodes, since typing something invalidates everything that comes after that location in the code, I don't think we need to delete tokens that are coming before our cursor (not %100 sure about this but feels like as i type something, only the idents coming after the cursor will change position) So we can further decrease the amount of tokens we need to kill after each keystroke. We would look for the region our cursor in, and kill anything that is after that region. This part assumes that we can fill in the gaps in the map, without re-type checking all file which I don't believe possible with the compiler right now.

TL;DR: We can introduce an indirection to how we store the tokens, so that rather than actually deleting individual tokens from map, we can mark all of them invalid in a single operation. We can also try to see if we can go below the file as the smallest unit we can deal with.
